File: estimators/est_quad1Diter.m
function xest=est_quad1Diter(photonsi,L,iter,eps)
%Eilers 2.64, k=1
if nargin<4
    eps=.1; %nm
end
if nargin<3
    iter=15;
end
if length(photonsi)==2
    itfun=@iteration;
else
    itfun=@iterationcenter;
end
pi=photonsi/sum(photonsi);
xest=0;
for k=1:iter
    xo=xest;
    xest=itfun(pi,xest,L);
    if sum((xest-xo).^2)<eps^2
        break
    end
end 
% xest=max(min(xest,L),-L); %avoid crazy high numbers
xest(abs(xest)>1.3*L)=NaN;
end

function dro=iteration(p,x0,L)
d12=p(1)-p(2);
dr=((L^2 + 4 *x0^2)* (-4 *L *x0 + d12* (L^2 + 4 *x0^2)))/(4 *(L^3 - 4 *L *x0^2));
dro=dr+x0;
end

function dro=iterationcenter(p,x0,L)
%p(1) at -L/2, p(2) at L/2, p(3) at 0
d12=p(1)-p(2);
pt=p(1)+p(2)-2*p(3);
dr=((L^2 + 6 *x0^2) *(-L^3 *(3 + pt) *x0 - 6 *L *(-4 + pt) *x0^3 + ...
   d12 *(L^4 - 36 *x0^4)))/(4 *(L^5 - 9 *L^3 *x0^2 + 36 *L *x0^4));
dro=dr+x0;
end

File: python/estimators/est_quad1Diter.py
import numpy as np

def est_quad1Diter(photonsi, L, iter=15, eps=0.1):
    """
    Eilers 2.64, k=1
    """

    if len(photonsi) == 2:
        itfun = iteration
    else:
        itfun = iterationcenter

    pi = photonsi / np.sum(photonsi)
    xest = np.array([0.0])
    for _ in range(iter):
        xo = xest.copy()
        xest = itfun(pi, xest, L)
        if np.sum((xest - xo)**2) < eps**2:
            break
    # xest = np.clip(xest, -L, L) # avoid crazy high numbers
    xest[np.abs(xest)>1.3*L] = np.nan
    return xest

def iteration(p, x0, L):
    d12 = p[0] - p[1]
    dr = ((L**2 + 4 * x0**2) * (-4 * L * x0 + d12 * (L**2 + 4 * x0**2))) / (4 * (L**3 - 4 * L * x0**2))
    dro = dr + x0
    return dro

def iterationcenter(p, x0, L):
    # p[0] at -L/2, p[1] at L/2, p[2] at 0
    d12 = p[0] - p[1]
    pt = p[0] + p[1] - 2 * p[2]
    dr = ((L**2 + 6 * x0**2) * (-L**3 * (3 + pt) * x0 - 6 * L * (-4 + pt) * x0**3 + 
           d12 * (L**4 - 36 * x0**4))) / (4 * (L**5 - 9 * L**3 * x0**2 + 36 * L * x0**4))
    dro = dr + x0
    return dro
